### PR TITLE
Add `on_model_save` callback with Python example to Callbacks docs

### PR DESCRIPTION
**Pull request by RizwanMunawar**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improved documentation to show how to access model metrics using the `on_model_save` callback during training. 📈

### 📊 Key Changes
- Added a new example in the callbacks documentation demonstrating how to retrieve and print training metrics (like best fitness, total loss, and other details) after each model checkpoint is saved.

### 🎯 Purpose & Impact
- Makes it easier for users to monitor and log training progress by accessing key metrics at each checkpoint.
- Helps both beginners and advanced users better understand and utilize callbacks for custom training workflows.
- Enhances transparency and control over the training process, leading to more informed model development. 🚀

### CODE REVIEW

**Comment by RizwanMunawar** [comment]
@glenn-jocher I notice a [user requested](https://github.com/ultralytics/ultralytics/issues/20524) the validation per epoch feature with Callbacks, and in the documentation, we don't have any information highlighting it. 

This PR will add the related Python example with `coco8.yaml`, and also the `on_model_save` callback is missing in the documentation, added here as well. Thanks

### Python example

```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")

batch, imgsz, workers, epochs, data, device = 2, 640, 1, 3, "coco8.yaml", "0"

def on_train_epoch_end(trainer):
    from pathlib import Path
    val_model = YOLO(Path(trainer.save_dir) / "weights" / "best.pt")

    print(f"\nRunning Validation {trainer.epoch+1}...")
    validation_results = val_model.val(
        data=data,
        imgsz=imgsz,
        batch=batch * 2,  # validation with 2x batch required
        workers=workers,
        device=device
    )
    # print(validation_results)

if __name__ == "__main__":
    # Add on_train_epoch_end callback.
    model.add_callback("on_model_save", on_train_epoch_end)

    # Run model training on the custom dataset.
    results = model.train(data=data, epochs=epochs, batch=batch,
                          imgsz=imgsz, device=device, workers=workers)
```

**Comment by RizwanMunawar** [comment]
@glenn-jocher, yes, I got your point. Based on my testing. it seems like we don't run validation every epoch, so only last epoch we are using validation.

```
Plotting labels to runs/detect/train/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 2 dataloader workers
Logging results to runs/detect/train
Starting training for 10 epochs...
Closing dataloader mosaic
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       1/10      4.35G      1.394      4.212       1.26         28        640: 100%|██████████| 28/28 [00:17<00:00,  1.63it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.13it/s]                   all        223        241    0.00193      0.536     0.0725     0.0396


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       2/10      5.13G      1.174      2.855       1.08         29        640: 100%|██████████| 28/28 [00:13<00:00,  2.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:02<00:00,  1.55it/s]                   all        223        241     0.0614      0.166     0.0643     0.0353


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       3/10      5.15G      1.153      2.438        1.1         28        640: 100%|██████████| 28/28 [00:12<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.09it/s]                   all        223        241      0.369      0.349      0.299      0.193


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       4/10      5.16G       1.14      2.153      1.085         32        640: 100%|██████████| 28/28 [00:12<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:02<00:00,  1.52it/s]                   all        223        241      0.423      0.673      0.417       0.28


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       5/10      5.17G      1.067      1.866      1.041         30        640: 100%|██████████| 28/28 [00:12<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.16it/s]                   all        223        241      0.415      0.638      0.402      0.266


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       6/10      5.19G      1.064      1.702      1.042         32        640: 100%|██████████| 28/28 [00:12<00:00,  2.18it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:02<00:00,  1.81it/s]                   all        223        241      0.442      0.827      0.489      0.338


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       7/10       5.2G      1.001      1.577      1.019         32        640: 100%|██████████| 28/28 [00:12<00:00,  2.18it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:03<00:00,  1.25it/s]                   all        223        241      0.456      0.816      0.474      0.337


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       8/10      5.22G     0.9725      1.457      1.015         29        640: 100%|██████████| 28/28 [00:12<00:00,  2.24it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:02<00:00,  1.57it/s]                   all        223        241      0.439      0.856      0.463      0.325


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       9/10      5.24G      0.904      1.377     0.9821         29        640: 100%|██████████| 28/28 [00:12<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:02<00:00,  1.71it/s]                   all        223        241      0.458      0.853      0.473      0.333


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      10/10      5.25G     0.9007      1.326     0.9755         28        640: 100%|██████████| 28/28 [00:12<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:02<00:00,  1.62it/s]                   all        223        241      0.462      0.866      0.476      0.346


10 epochs completed in 0.046 hours.
Optimizer stripped from runs/detect/train/weights/last.pt, 5.5MB
Optimizer stripped from runs/detect/train/weights/best.pt, 5.5MB

Validating runs/detect/train/weights/best.pt...
Ultralytics 8.3.128 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
YOLO11n summary (fused): 100 layers, 2,582,542 parameters, 0 gradients, 6.3 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:05<00:00,  1.28s/it]
                   all        223        241      0.461      0.869      0.476      0.346
              negative        142        154       0.58      0.826        0.6      0.442
              positive         81         87      0.342      0.913      0.352      0.251
Speed: 0.4ms preprocess, 2.6ms inference, 0.0ms loss, 6.5ms postprocess per image
Results saved to runs/detect/train
```

**Comment by glenn-jocher** [comment]
@RizwanMunawar yes of course we run validation after every epoch. I don't understand what you mean. How else would get training curves with metrics per epoch?

Yes anything else here would work like sending a notice or saving a file etc.

Screenshot 2025-05-07 at 20 46 18

**Comment by RizwanMunawar** [comment]
@glenn-jocher Ok, I adjusted it with the notification example. It may not be the best use case here, but it's a good starting point. What do you think? Thanks

```python
from ultralytics import YOLO
from ultralytics.utils.checks import check_requirements

check_requirements("plyer")

model = YOLO("yolo11n.pt")


def on_model_save(model):
    """Send notification once after each iteration model checkpoint saved."""
    from plyer import notification

    notification.notify(
        title="Ultralytics YOLO 🚀",
        message=f"Epoch {model.epoch + 1} is completed and model saved in the directory: {model.save_dir}",
        timeout=5,  # seconds
    )


if __name__ == "__main__":
    # Add on_train_epoch_end callback.
    model.add_callback("on_model_save", on_model_save)

    # Run model training on custom dataset.
    results = model.train(data="coco8.yaml", epochs=3, workers=1)```

**Comment by RizwanMunawar** [comment]
@glenn-jocher Ok, got it, I've updated the example to display key model metrics such as `best_fitness`, `tloss`, and others. It might not be the ideal use case in this context, but I think it serves as a good starting point. Please let me know your thoughts. Thanks.

```python
from ultralytics import YOLO
from ultralytics.utils import LOGGER

model = YOLO("yolo11n.pt")


def on_model_save(model):
    """Print model details after each epoch."""
    LOGGER.info(
        f"Model details\n"
        f"Best fitness: {model.best_fitness}, "
        f"Loss names: {model.loss_names}, "  # List of loss names
        f"Metrics: {model.metrics}, "
        f"Total loss: {model.tloss}"  # Total loss value
    )  


if __name__ == "__main__":
    # Add on_train_epoch_end callback.
    model.add_callback("on_model_save", on_model_save)

    # Run model training on custom dataset.
    results = model.train(data="coco8.yaml", epochs=3, workers=1)
```

**Comment by glenn-jocher** [comment]
@RizwanMunawar I've fixed the example, on_model_save is a trainer callback, it's already shown in the correct trainer callback section, there's no such thing as a model callback. 

I've also fixed the argument passed by renaming it from model to trainer.

**Comment by glenn-jocher** [comment]
Also the text is incorrect, this callback is not run every epoch, it is only run when a checkpoint is saved.

**Comment by UltralyticsAssistant** [comment]
Fantastic work, @RizwanMunawar and @glenn-jocher! 🚀 Your improvements to the callbacks documentation will empower users to track and understand their training progress with greater clarity. As Benjamin Franklin said, "An investment in knowledge always pays the best interest." Your contributions make Ultralytics HUB and the YOLO community more accessible and insightful for everyone. Thank you for your dedication and for helping others unlock the full potential of model training!
